Fixing L4 offset for IP packets with Options field

evershalik opened the pull request:
This PR fixed #30 

**Problem Statement:**
The current implementation of the codebase does not handle IP and TCP packets with a header length greater than 20 bytes correctly. This issue is particularly problematic for packets containing options, which increase the header length beyond 20 bytes.

**Approach:**
To address this issue, we have taken the following approaches:

*Firstly*, I have modified the code to use the `l4_offset` variable(IP Header Length) instead of `sizeof(iphdr struct)` when adding to the iph pointer. This ensures that the code correctly traverses to the end of the IP header( or the start of the transport layer header), even in cases where the header length is greater than 20 bytes.

*Secondly*, I have modified the code to use the `data_offset` variable(TCP Header Length) instead of `sizeof(tcphdr struct)` when adding to the tcph pointer. This ensures that the code correctly traverses to the end of the TCP header (or start of the payload), even in cases where the header length is greater than 20 bytes.

By implementing these changes, I can ensure that IP and TCP packets with options are handled correctly, which improves the reliability and stability of the system.

aka320 commented on connection-limit/connection_limit_kern.c:
```diff
@@ -273,17 +273,20 @@ int _xdp_limit_conn(struct xdp_md *ctx)
+    
     struct iphdr *iph = (struct iphdr *)(data + sizeof(struct ethhdr));
-    if (iph + 1 > data_end)
+    __u8 l4_offset = iph->ihl * 4; // ip header length
```
Minimum value of IHL is 5. Add a check to discard invalid IP packets with IHL less than 5.

aka320 commented on connection-limit/connection_limit_kern.c:
```diff
@@ -273,17 +273,20 @@ int _xdp_limit_conn(struct xdp_md *ctx)
     /* Check if its valid tcp packet */
-    struct tcphdr *tcph = (struct tcphdr *)(iph + 1);
+    struct tcphdr *tcph = (struct tcphdr *)((unsigned char *)iph + l4_offset);
     if (tcph + 1 > data_end)
```
TCP header is also a variable-length header, which is determined by Data Offset field. Apply the same IHL logic here.

aka320 commented on connection-limit/connection_limit_kern.c:
```diff
@@ -273,17 +273,20 @@ int _xdp_limit_conn(struct xdp_md *ctx)
+    
     struct iphdr *iph = (struct iphdr *)(data + sizeof(struct ethhdr));
-    if (iph + 1 > data_end)
+    __u8 l4_offset = iph->ihl * 4; // ip header length
```
Endianness is handled in the definition of 'struct iphdr'. However please have a test-case to verify that the value for IHL is read correctly from the packet.

dthaler replied:
What sort of check did you have in mind?  I am not sure one is necessary.

aka320 commented on ipfix-flow-exporter/bpf_ipfix_egress_kern.c:
```diff
@@ -249,13 +249,21 @@ void parse_ipv4(struct __sk_buff *skb, u64 l3_offset)
+
+    //  Check if there are any options present
+    if (l4_offset > sizeof(struct iphdr *)) 
+        thdr = (struct iphdr *)((unsigned char *)iph + l4_offset);
```
No need to have if condition for this. Just initialize thdr to ((unsigned char *)iph + l4_offset), which will work for all the cases.

aka320 replied:
thdr is not of type (struct iphdr *) It is the start of Transport header which is either TCP or UDP. You can define a void pointer as parse_icmp_type() & parse_port() functions accept void pointer as arg.

aka320 reviewed (changes requested):
1. The comments given are applicable for all the eBPF applications. Please fix in all files even though its not explicitly commented in that specific file.

2. Please include the tests results (different values of IHL including negative cases like invalid IHL value)

3. Add a Commit log message with a short description of RootCause (pointing to the source Issue) & Details of the Fix

evershalik commented:
> 

I have gone ahead and fixed the issue in all relevant files, as you suggested, even if the specific comment was not explicitly mentioned in those files.
I have added a commit log message that provides a short description of the root cause of the issue (as pointed out in the source issue), along with details of the fix that has been implemented. This should help with tracking the changes and understanding the reasons for the changes made.
Regarding the test results, it may take some time to complete these tests and ensure that everything is working as expected.
Thank you for your feedback.

aka320 commented:
1. You can remove the diagrams in the Description. Text is enough.
2. Next time when you address review comments, do not overwrite the previous commit. It doesnt give option to review the additional changes made for the review comments.

evershalik commented:
Thank you for your feedback.
I have removed the diagrams in the Description.
And I apologize for overwriting the previous commit. I will take care of this in the future.

aka320 commented:
Changes look good. Please update once testing is done.

dthaler commented on connection-limit/connection_limit_kern.c:
```diff
@@ -273,18 +273,23 @@ int _xdp_limit_conn(struct xdp_md *ctx)
 
-    /* Check if its valid ip packet */
+    
     struct iphdr *iph = (struct iphdr *)(data + sizeof(struct ethhdr));
```
Looks like a bug, is there someplace that handles IPv6?

aka320 replied:
@dthaler You are right. There is no IPv6 handling as of now. Check for IPv6 packets is missing in connection-limiting, which is present in rate-limiting program. Below is the snippet from _xdp_ratelimit() where they are ignoring other than IPv4 packets.
```
    uint16_t eth_type = eth->h_proto;
    if (ntohs(eth_type) != ETH_P_IP) {
        return XDP_PASS;
    }
```

dthaler replied:
Yes, that's what is needed here until this supports IPv6.  I've marked this PR as "changes requested" until that code is added here.

dthaler reviewed (changes requested):
This code isn't safe because it doesn't check whether the packet is IPv4 before trying to interpret the IPv4 header.

dthaler commented on ipfix-flow-exporter/bpf_ipfix_egress_kern.c:
```diff
@@ -249,13 +252,18 @@ void parse_ipv4(struct __sk_buff *skb, u64 l3_offset)
-        return;
+    u8 l4_offset = iph->ihl * 4; // ip header length
 
+    /* Check if its valid ip packet */
```
```suggestion
    /* Check if it's a valid IPv4 packet */
```

dthaler commented on ipfix-flow-exporter/bpf_ipfix_ingress_kern.c:
```diff
@@ -247,13 +250,18 @@ void parse_ipv4(struct __sk_buff *skb, u64 l3_offset)
-        return;
+    u8 l4_offset = iph->ihl * 4; // ip header length
 
+    /* Check if its valid ip packet */
```
```suggestion
    /* Check if it's a valid IPv4 packet */
```

dthaler commented on ratelimiting/ratelimiting_kern.c:
```diff
@@ -97,18 +97,22 @@ static __always_inline int _xdp_ratelimit(struct xdp_md *ctx)
+    struct tcphdr *tcph = (struct tcphdr *)((unsigned char *)iph + l4_offset);
+    __u16 data_offset = tcph->doff * 4; // tcp header length
+
     /* Check if its valid tcp packet */
```
```suggestion
    /* Check if it's a valid TCP packet */
```

dthaler commented on .github/workflows/ci-build-traffic-mirroring.yaml:
```diff
@@ -0,0 +1,72 @@
+          make defconfig
+          
+      - name: Clone kernel function repository
+        uses: actions/checkout@v2
```
```suggestion
        uses: actions/checkout@b0e28b5ac45a892f91e7d036f8200cf5ed489415
```
use hash instead of version.  I see some programs use version and others use hash, we should be consistent at always using hash.

dthaler commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,104 @@
+# Copyright Contributors to the L3AF Project.
+# SPDX-License-Identifier: (GPL-2.0 OR BSD-2-Clause)
```
is there a reason the codeql file needs to have GPL in the SPDX?

dthaler commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,104 @@
+
+    # Initializes the CodeQL tools for scanning.
+    - name: Initialize CodeQL
+      uses: github/codeql-action/init@v2
```
```suggestion
      uses: github/codeql-action/init@04df1262e6247151b5ac09cd2c303ac36ad3f62b
```

dthaler commented on .github/workflows/codeql.yml:
```diff
@@ -0,0 +1,104 @@
+        make defconfig
+
+    - name: Clone eBPF-Package-Repository
+      uses: actions/checkout@v2
```
```suggestion
      uses: actions/checkout@b0e28b5ac45a892f91e7d036f8200cf5ed489415
```

dthaler commented on buildscript.sh:
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
+# Copyright Contributors to the L3AF Project.
+# SPDX-License-Identifier: (GPL-2.0 OR BSD-2-Clause)
```
Any reason GPL is needed in the build script SPDX for all programs?

dthaler commented on connection-limit/connection_limit_kern.c:
```diff
@@ -327,4 +339,4 @@ int _xdp_limit_conn(struct xdp_md *ctx)
 }
 
-char _license[] SEC("license") = "GPL";
+char _license[] SEC("license") = "Dual BSD/GPL";
```
According to git, Santhosh was the author of this file, so only Santhosh can change the license.  Hence we need Santhosh to approve this PR before merging it.  This does seem unrelated to "fixing L4 offset for IP packets with Options field" so I would recommend all license related changes in this PR be moved to a separate PR.

dthaler commented on traffic-mirroring/README.md:
```diff
@@ -0,0 +1,113 @@
+Traffic Mirroring Tool (TMT)
```
What does this addition have to do with "Fixing L4 offset for IP packets with Options field"?
Maybe move this to a separate PR?  or else change the PR title?

dthaler commented on traffic-mirroring/README.md:
```diff
@@ -0,0 +1,113 @@
+\--iface=\<interface name\> \\  
+\--direction=ingress \\  
+\--map-name=\<location of the tc-root map for chaining\> \\  
+\--src-address=\<list of the source addresses (Use 0.0.0.0 for mirroring packets with "any" src IP)\> \\  
```
```suggestion
\--src-address=\<list of the source addresses (Use 0.0.0.0 for mirroring packets with "any" IPv4 src)\> \\  
```

dthaler commented on traffic-mirroring/README.md:
```diff
@@ -0,0 +1,113 @@
+\--iface=\<interface name\> \\  
+\--direction=egress \\  
+\--map-name=\<location of the tc-root map for chaining\> \\  
+\--dst-address=\<list of the destination addresses (Use 0.0.0.0 for mirroring packets with "any" dst IP)\> \\  
```
```suggestion
\--dst-address=\<list of the destination addresses (Use 0.0.0.0 for mirroring packets with "any" IPv4 dst)\> \\  
```

dthaler commented on traffic-mirroring/mirroring_egress_kern.c:
```diff
@@ -0,0 +1,269 @@
+    __u8 data[4];
+};
+
+/* TODO: Describe what this PIN_GLOBAL_NS value 2 means???
```
Can this TODO be done before code gets merged?

dthaler commented on traffic-mirroring/mirroring_egress_kern.c:
```diff
@@ -0,0 +1,269 @@
+    int *ifany;
+    int iface_key = 1;
+
+    /* Lookup what ifindex to redirect packets to */
```
```suggestion
    /* Look up what ifindex to redirect packets to */
```

dthaler commented on traffic-mirroring/mirroring_egress_kern.c:
```diff
@@ -0,0 +1,269 @@
+    struct iphdr *iph = (struct iphdr *)(data + l3_off);
+
+    if (iph->protocol == IPPROTO_TCP) {
+        l7_off = l4_off + sizeof(struct tcphdr); // L7 (e.g. HTTP) header offset
+    } else if (iph->protocol == IPPROTO_UDP) {
+        l7_off = l4_off + sizeof(struct udphdr); // L7 (e.g. HTTP) header offset
+    } else if (iph->protocol == IPPROTO_ICMP) {
+        l7_off =
+            l4_off + sizeof(struct icmphdr); // L7 (e.g. HTTP) header offset
```
```suggestion
        l7_off = l4_off + sizeof(struct tcphdr); // L7 (e.g., HTTP) header offset
    } else if (iph->protocol == IPPROTO_UDP) {
        l7_off = l4_off + sizeof(struct udphdr); // L7 (e.g., HTTP) header offset
    } else if (iph->protocol == IPPROTO_ICMP) {
        l7_off =
            l4_off + sizeof(struct icmphdr); // L7 (e.g., HTTP) header offset
```

dthaler commented on traffic-mirroring/mirroring_egress_kern.c:
```diff
@@ -0,0 +1,269 @@
+
+    // Binary representation of ifany variable. Reading from LSB to MSB:
+    //      0th bit True: allow all/any IP addresses
+    //      1th bit True: allow all/any source port
+    //      2th bit True: allow all/any destination port
```
```suggestion
    //      1st bit True: allow all/any source port
    //      2nd bit True: allow all/any destination port
```

dthaler commented on traffic-mirroring/mirroring_egress_kern.c:
```diff
@@ -0,0 +1,269 @@
+        }
+    }
+
+    /* Check if the packet's destination ip falls in the remote endpoint CIDR */
```
```suggestion
    /* Check if the packet's destination IP address falls in the remote endpoint prefix */
```
Use IP version agnostic terminology where possible

dthaler commented on traffic-mirroring/mirroring_egress_kern.c:
```diff
@@ -0,0 +1,269 @@
+    //      2th bit True: allow all/any destination port
+    ifany = bpf_map_lookup_elem(&egress_any, &iface_key);
+    if (ifany) {
+        if (((*ifany) >> (0)) % 2 == 1) {
+            allow_all_ip = true;
+        }
+        if (((*ifany) >> (1)) % 2 == 1) {
+            allow_all_src_ports = true;
+        }
+        if (((*ifany) >> (2)) % 2 == 1) {
```
This would be easier to follow if there were `#define` values for bits that could be AND'ed, e.g.
```
if (((*ifany) & ALLOW_ALL_IP) {
```

dthaler commented on traffic-mirroring/mirroring_egress_kern.c:
```diff
@@ -0,0 +1,269 @@
+    /* Check if the packet's destination ip falls in the remote endpoint CIDR */
+    if (allow_all_ip == true || bpf_map_lookup_elem(&dst_address, &dkey)) {
+        uint32_t protocol = iph->protocol;
+        // 1 = ICMP
+        // 6 = TCP
+        // 17 = UDP
```
SHouldn't need these comments since the defines used below already have those numbers and one can look there.

dthaler commented on traffic-mirroring/mirroring_ingress_kern.c:
```diff
@@ -0,0 +1,270 @@
+    __u8 data[4];
+};
+
+/* TODO: Describe what this PIN_GLOBAL_NS value 2 means???
```
Can this TODO be done before approving the PR?

dthaler commented on traffic-mirroring/mirroring_ingress_kern.c:
```diff
@@ -0,0 +1,270 @@
+    int *ifany;
+    int iface_key = 1;
+
+    /* Lookup what ifindex to redirect packets to */
```
```suggestion
    /* Look up what ifindex to redirect packets to */
```

evershalik commented:
Sorry for this. I have synced the forked repository by mistake due to which commits made by other people are also showing in this PR. What should I do now to revert the this?
